astropy/io/misc/hdf5.py

```diff
@@ -150,7 +150,7 @@ def read_table_hdf5(input, path=None, character_as_bytes=True):
     # Create a Table object
     from astropy.table import Table, meta, serialize
 
-    table = Table(np.array(input))
+    table = Table(np.array(input, copy=COPY_IF_NEEDED))
```

Comment:
That this fails is really an upstream bug!  But if this works, it is good regardless, since `Table(...)` already makes a copy by default, so there is not much point in making another one for `np.array(...)`.

Reply:
Hard to tell when this started happening so I don't know how to decide if this is upstream bug or not.

Reply:
`np.array(input)` is the one that fails, because it calls `input.__array__(copy=...)` and the h5py implementation doesn't have that. In fact, it looks like @neutrinoceros already had a PR to fix this in h5py (https://github.com/h5py/h5py/pull/2416), but perhaps that is not available in the version we are using. Anyway,  as said, I think this is a reasonable idea regardless.

Though, one thing: in the PR, it states that it is impossible for `h5py` *not* to make a copy -- if so, perhaps we can add `copy=False` to the `Table` call? Let's ping @neutrinoceros for advice.

Reply:
You mean `copy=True`? Because `copy=False` is what throwing the error, I think.

Reply:
Oh wait, I guess these are 2 different problems:

From the log: `__array__ implementation doesn't accept a copy keyword, so passing copy=False failed. __array__ must implement 'dtype' and 'copy' keyword arguments.`

And from this I tried locally:

```python
>>> import numpy as np
>>> np.array([1,2,3], copy=False)
ValueError: Unable to avoid copy while creating an array as requested.
```

Never mind...

Reply:
> it looks like @neutrinoceros already had a PR to fix this in h5py (h5py/h5py#2416), but perhaps that is not available in the version we are using.

I confirm this PR hasn't been included in a release yet.
And yes, passing `copy=False` or `copy=COPY_IF_NEEDED` wouldn't work since h5py will unconditionally copy here.
My advice would be to add a filter for this deprecation warning in pyproject.toml and link to h5py/h5py#2416 to show that it's being handled.

I assume we'll be cutting a new h5py release soon anyway if only to add cp313 wheels, but I'll check with the rest of the team !

Reply:
Yeah, `np.array(h5py_instance)` always seems to return a copy, hence I wondered about adding `copy=False` to the table, as in `Table(np.array(input, copy=COPY_IF_NEEDED), copy=False)`. Obviously, would be clearer to use `Table(np.array(input, copy=True), copy=False)` -- but perhaps that doesn't work either?

Reply:
Alas, `table = Table(np.array(input, copy=True), copy=False)` does not fix the problem.

Reply:
I don't think we're even able to mitigate the issue on the astropy side: passing more arguments isn't going to help with signatures that don't support them. I think `Table(np.array(input, copy=True), copy=False)` in principle is correct but just makes the default behavior explicit, and is not a clear win to me. It's certainly fine to do it, it just won't make the `DeprecationWarning` go away.

Reply:
(or does it now ? I don't know if normal CI is expected to fail or not, and I haven't *checked* anything I said in passing here)

Reply:
It didn't help when I tested it locally. I modified this PR to just ignore the warning for now, as @neutrinoceros suggested and added him as co-author. 🤞